File: prisma/refs/py/serialize-and-deserialize-binary-tree.py
class Codec:
    def serialize(self, root):
        out = []

        def dfs(n):
            if not n:
                out.append("#")
                return
            out.append(str(n.val))
            dfs(n.left)
            dfs(n.right)

        dfs(root)
        return ",".join(out)

    def deserialize(self, data):
        tokens = iter(data.split(","))

        def build():
            t = next(tokens)
            if t == "#":
                return None
            node = TreeNode(int(t))  # noqa: F821 (injected by harness)
            node.left = build()
            node.right = build()
            return node

        return build()

File: prisma/refs/serialize-and-deserialize-binary-tree.ts
type TreeNode = { val: number; left: TreeNode | null; right: TreeNode | null };

export class Codec {
  serialize(root: TreeNode | null): string {
    const out: string[] = [];
    const dfs = (n: TreeNode | null) => {
      if (!n) {
        out.push("#");
        return;
      }
      out.push(String(n.val));
      dfs(n.left);
      dfs(n.right);
    };
    dfs(root);
    return out.join(",");
  }

  deserialize(data: string): TreeNode | null {
    const tokens = data.split(",");
    let i = 0;
    const build = (): TreeNode | null => {
      const t = tokens[i++];
      if (t === "#") return null;
      const node: TreeNode = { val: Number(t), left: null, right: null };
      node.left = build();
      node.right = build();
      return node;
    };
    return build();
  }
}
